In [1]:
# Import des données 

import pandas as pd 
file_path_train = 'C:/Users/HP PROBOOK/Documents/churn-bigml-80.csv' 
df_train = pd.read_csv(file_path_train)

df_train.head() 

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False

In [2]:
df_train.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2666 entries, 0 to 2665
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   State                   2666 non-null   object 
 1   Account length          2666 non-null   int64  
 2   Area code               2666 non-null   int64  
 3   International plan      2666 non-null   object 
 4   Voice mail plan         2666 non-null   object 
 5   Number vmail messages   2666 non-null   int64  
 6   Total day minutes       2666 non-null   float64
 7   Total day calls         2666 non-null   int64  
 8   Total day charge        2666 non-null   float64
 9   Total eve minutes       2666 non-null   float64
 10  Total eve calls         2666 non-null   int64  
 11  Total eve charge        2666 non-null   float64
 12  Total night minutes     2666 non-null   float64
 13  Total night calls       2666 non-null   int64  
 14  Total night charge      2666 non-null   

In [3]:
# Séparation des variables 

categorical_features = ['State', 'International plan', 'Voice mail plan', 'Area code'] 
numeric_features = ['Account length', 'Number vmail messages', 'Total day minutes', 'Total day calls', 'Total day charge', 
                    'Total eve minutes', 'Total eve calls', 'Total eve charge', 'Total night minutes', 'Total night calls', 'Total night charge', 
                    'Total intl minutes', 'Total intl calls', 'Total intl charge', 'Customer service calls'] 

# Encodage et standardisation 

from sklearn.preprocessing import OneHotEncoder, StandardScaler 
from sklearn.compose import ColumnTransformer 

# Application 
preprocessor = ColumnTransformer(
    transformers=[
    ('num', StandardScaler(), numeric_features), 
    ('cat', OneHotEncoder(), categorical_features)
    ])

In [4]:
# Application du prétraitement aux données de train

# Création d'un nouveau Dataframe pour les caratérisques sans la cible 
X_train = df_train.drop(columns=["Churn"]) 

# Extraction de la colonne cible 
y_train = df_train['Churn'] 

# Application de la transformation 
X_train = preprocessor.fit_transform(X_train) 


In [5]:
# Affichage des dimensions après prétraitement 

print(X_train.shape) 
print(y_train.shape) 

(2666, 73)
(2666,)


In [6]:
# Chargeons les données de test 
file_path_test = 'C:/Users/HP PROBOOK/Documents/churn-bigml-20.csv' 

df_test = pd.read_csv(file_path_test) 

df_test.head() 

State  Account length  Area code International plan Voice mail plan  \
0    LA             117        408                 No              No   
1    IN              65        415                 No              No   
2    NY             161        415                 No              No   
3    SC             111        415                 No              No   
4    HI              49        510                 No              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                      0              184.5               97   
1                      0              129.1              137   
2                      0              332.9               67   
3                      0              110.4              103   
4                      0              119.3              117   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             31.37              351.6               80             29.89   
1             21.95              228.5               83             19.42   
2             56.59              317.8               97             27.01   
3             18.77              137.3              102             11.67   
4             20.28              215.1              109             18.28   

   Total night minutes  Total night calls  Total night charge  \
0                215.8                 90                9.71   
1                208.8                111                9.40   
2                160.6                128                7.23   
3                189.6                105                8.53   
4                178.7                 90                8.04   

   Total intl minutes  Total intl calls  Total intl charge  \
0                 8.7                 4               2.35   
1                12.7                 6               3.43   
2                 5.4                 9               1.46   
3                 7.7                 6               2.08   
4                11.1                 1               3.00   

   Customer service calls  Churn  
0                       1  False  
1                       4   True  
2                       4   True  
3                       2  False  
4                       1  False

In [7]:
# Verification des dimensions 
print(f'les dimensions du jeu de données de test sont : {df_test.shape}') 

les dimensions du jeu de données de test sont : (667, 20)


In [8]:
# Prétraitement des données de test 

# Création d'un nouveau Dataframe pour les caratérisques sans la cible 
X_test = df_test.drop(columns=["Churn"]) 

# Extraction de la colonne cible 
y_test = df_test['Churn'] 

# Application de la transformation 
X_test = preprocessor.fit_transform(X_test) 


In [9]:
# Affichage des dimensions après prétraitement 

print(X_test.shape) 
print(y_test.shape) 

(667, 73)
(667,)


In [10]:
# Implémentation et entraînement des modèles 

from sklearn.linear_model import LogisticRegression 
from sklearn.neighbors import KNeighborsClassifier 
from sklearn.svm import SVC 
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier 
from sklearn.metrics import roc_auc_score, classification_report 

In [11]:
# Definir les modèles à entraîner 

models = {
'Regression Logistique' : LogisticRegression(random_state=42),
'Suport Vecteur Machine' : SVC(probability=True, random_state=42),
'K-N' : KNeighborsClassifier(), 
'GBoost' : GradientBoostingClassifier(random_state=42), 
'Forêt Aléatoire': RandomForestClassifier(random_state=42)
}
   

In [12]:
# Entraînement et Evaluation des modèles 

for name, model in models.items(): 
    model.fit(X_train, y_train) 
    y_pred = model.predict(X_test) 
    auc_score = roc_auc_score(y_test, y_pred) 
    print(f"Model: {name}") 
    print(classification_report(y_test, y_pred)) 
    print(f"AUC: {auc_score}\n")  

Model: Regression Logistique
              precision    recall  f1-score   support

       False       0.88      0.96      0.92       572
        True       0.49      0.21      0.29        95

    accuracy                           0.86       667
   macro avg       0.68      0.59      0.61       667
weighted avg       0.82      0.86      0.83       667

AUC: 0.5869065145380935

Model: Suport Vecteur Machine
              precision    recall  f1-score   support

       False       0.91      0.99      0.95       572
        True       0.85      0.43      0.57        95

    accuracy                           0.91       667
   macro avg       0.88      0.71      0.76       667
weighted avg       0.90      0.91      0.90       667

AUC: 0.7096705925653295

Model: K-N
              precision    recall  f1-score   support

       False       0.89      0.99      0.94       572
        True       0.86      0.25      0.39        95

    accuracy                           0.89       667
   macro

Analyse et conclusion 

- Regression Logistique : Ce modèle a une bonne précision pour les non-churners (False) mais une faible performance pour les churners (True). L'AUC est relativement faible, indiquant une performance globale moyenne. 

- SVM : Le SVM a montré une bonne performance avec une précision élevée pour les deux classes, mais le rappel pour les churners est encore modéré. L'AUC de 0.7097 indique une performance acceptable. 

- KNN : KNN a une bonne précision, mais le rappel pour les churners est faible, ce qui affecte le F1-score. L'AUC de 0.6228 montre une performance globale moyenne. 

- GBoot : Le modèle Gradient Boosting a des performances très solides avec une haute précision et un bon rappel pour les churners. L'AUC de 0.8333 indique une excellente performance globale. 

- Forêt Aléatoire : Le modèle de Forêt Aléatoire a une excellente précision pour les deux classes, mais le rappel pour les churners est légèrement inférieur comparé au Gradient Boosting. L'AUC de 0.7939 montre une très bonne performance globale. 


Conclusion
Parmi les modèles testés, le Gradient Boosting a montré les meilleures performances globales avec une précision et un rappel élevés pour les churners, ainsi qu'un excellent score AUC. Ce modèle est donc recommandé pour prédire le churn dans ce contexte. 